Lib/unittest/case.py

```diff
@@ -153,6 +154,35 @@ def _is_subtype(expected, basetype):
         return all(_is_subtype(e, basetype) for e in expected)
     return isinstance(expected, type) and issubclass(expected, basetype)
 
+
+def _heuristic_diff(a: list[str], b: list[str]) -> Iterator[str]:
+    """After testing the magnitude of the inputs, preferably return the output
+    of difflib.ndiff, but fallback to difflib.unified_diff for prohibitively
+    expensive inputs. How cost is calclated:
+
+        cost = (number of differing lines
+                * total length of all differing lines)
+
+    See bpo-19217 for context.
+    """
+    cost_limit = 1_000_000
+
+    # unified diff is always cheap, so we can use it to measure the magnitude
+    # of the differences, which is a proxy for the cost of difflib.ndiff
+    udiff = list(difflib.unified_diff(a, b, fromfile="expected", tofile="got"))
+    udiff_differing_lines = [l for l in udiff
+                             if l.startswith('-') or l.startswith('+')]
+    num_difflines = len(udiff_differing_lines)
+    total_diffline_length = sum(len(l) for l in udiff_differing_lines)
+
+    diff_cost = num_difflines * total_diffline_length
```

Comment:
I don't have a CS degree, and don't have a strong understanding of time complexity, which is why I'm not sure if I can explain why it works correctly, so I will try, but please check my work if you can :)

`unified_diff` calls `difflib.SequenceMatcher.get_opcodes`, and then transparently converts those opcodes into a diff. It looks to me like the time complexity is `O(a * b)`, given two strings, `a`, and `b` to compare. So, diffs are produced in linear time by comparing one line at a time.

`ndiff`, on the other hand, uses many of the higher-level features found in `difflib.Differ`. In particular, `difflib.Differ._fancy_replace` is probably where the time complexity balloons. It is documented as follows:

```
When replacing one block of lines with another, search the blocks
for *similar* lines; the best-matching pair (if any) is used as a
synch point, and intraline difference marking is done on the
similar pair. Lots of work, but often worth it.
```

Let's look at the time complexity of each frame independently, imagining that the functions it calls have zero-cost, so we can then sum up these complexities later.

| Func or Method | Time Complexity | Notes |
| -- | -- | -- |
| `difflib.ndiff` | `O(1)` | proxy to `difflib.Differ.compare` |
| `difflib.Differ.compare` | `O(a + b)` (ignoring calls inside the loop, which will be summed later) | `cruncher.get_opcodes` is O(a + b), and the size of the opcodes is linearly related to the extent of the diff, I believe |
| `difflib.Differ.dump` | `O(n)` | called for opcodes `delete`, `insert`, and `equal` |
| `difflib.Differ._fancy_replace` | `O(???)` –– a lot | I count at least 3 nested for loops in some edge cases. There is caching to take into account. I really don't have a clue – let's call it `O((a + b)^2)` on a good day. This is only called for `replace` opcodes, where it annotates the replacement for a better diff. Very pricy.

So, `ndiff` has an overall time complexity of:

```
O(a + b)    :: from calls to `get_opcodes`, and looping over those opcodes
* O(n)      :: this is the effect of calls to `Differ._dump` inside the loop in `Differ.compare`,
               which can effectively be ignored because it's not worse than `_fancy_replace`
* O((a + b)^2)    :: ...or more. This is the big nasty – calls to `_fancy_replace` inside the loop
               in `Differ.compare`
```

Therefore, the time complexity of `ndiff` is in the neighborhood of `O(n^2)`, I think? `unified_diff` seems closer to O(n).

I might be totally wrong, though... @JelleZijlstra what do you think?

Reply:
I did try to improve documentation of the source in b9f2f9db79d971a2022ad915d42198827aed7b18 based on my limited understanding.

Reply:
I did try to improve documentation of the source in b9f2f9db79d971a2022ad915d42198827aed7b18 based on my limited understanding.

Reply:
Thanks, I think the text you added and the discussion here are very helpful! From a little IPython experimentation it does look like unified_diff is linear and ndiff is quadratic on a simple task (comparing a list of N "a" with N "b"):
```
In [19]: def nd(n):
    ...:     a = ["a"] * n
    ...:     b = ["b"] * n
    ...:     list(difflib.ndiff(a, b))
    ...: 
    ...: 

In [20]: def ud(n):
    ...:     a = ["a"] * n
    ...:     b = ["b"] * n
    ...:     list(difflib.unified_diff(a, b))
    ...: 
    ...: 

In [21]: %timeit ud(1000)
620 µs ± 12.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [24]: %timeit ud(10000)
6.96 ms ± 513 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [25]: %timeit ud(100000)
74.5 ms ± 9.38 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [26]: %timeit nd(100)
19.9 ms ± 638 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [27]: %timeit nd(1000)
1.89 s ± 135 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [28]: %timeit nd(10000)
```
For unified_diff, 10x as much data makes it ~10x slower, but for ndiff 10x the data makes it ~100x slower. I am not sure I am going to wait for that last one to finish.

Reply:
Ok cool, so it seems like what I added in b9f2f9d is accurate, then – I'm glad to hear it!

Do you think that b9f2f9d satisfies what you are looking for in terms of better documenting the heuristic?

Reply:
Yes, this is good!